lms/services/organization.py

```diff
@@ -0,0 +1,78 @@
+from logging import getLogger
+from typing import List, Optional
+
+import sqlalchemy as sa
+from sqlalchemy.orm import Session
+
+from lms.models import ApplicationInstance, GroupInfo, Organization
+
+LOG = getLogger(__name__)
+
+
+class OrganizationService:
+    """A service for dealing with organization actions."""
+
+    def __init__(self, db_session: Session):
+        self._db_session = db_session
+
+    def get_by_linked_guid(self, guid) -> List[Organization]:
+        """Get organizations which match the provided GUID."""
+
+        # There are huge numbers of null GUIDs, don't match on them
+        if not guid:
+            return []
+
+        return (
+            self._db_session.query(Organization)
+            .join(ApplicationInstance)
+            .outerjoin(GroupInfo)
+            .filter(
+                sa.or_(
+                    ApplicationInstance.tool_consumer_instance_guid == guid,
+                    GroupInfo.tool_consumer_instance_guid == guid,
+                )
+            )
+            .all()
+        )
+
+    def auto_assign_organization(
+        self, application_instance: ApplicationInstance
+    ) -> Optional[Organization]:
+        """
+        Automatically associate an application instance with an org by GUID.
+
+        If there is no GUID matching is skipped, and we return `None`. If no
+        match can be found, then a new organization is created.
+        """
+
+        # We can't match by GUID if there isn't one
+        guid = application_instance.tool_consumer_instance_guid
+        if not guid:
+            return None
```

Comment:
This should not happen on the context of a launch but it's good the function guards against it anyway.